fix empty key_name on aws_spot_fleet_request

svenwltr opened the pull request:
This fixes one part of https://github.com/terraform-providers/terraform-provider-aws/issues/492.

Basically Terraform sends an empty string for the `key_name` values to the AWS API, if the value isn't set in the template. This causes the instance creation to fail:

```
Instance fail to start with failed: Invalid value '' for keyPairNames. It should not be blank.
```

I am not sure, if it is worth doing an acceptance test here. It seems a bit complicated, because it would require to wait for the spot fleet instance creation. I am not sure how to handle this in a acceptance test (maybe with timeouts?). Please tell me, if I should add an test here.

radeksimko commented:
Hi @svenwltr 
thanks for this patch!

> I am not sure, if it is worth doing an acceptance test here. It seems a bit complicated, because it would require to wait for the spot fleet instance creation. I am not sure how to handle this in a acceptance test (maybe with timeouts?). Please tell me, if I should add an test here.

It would be better to add a test covering this bugfix. We [already wait for fulfilment](https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_spot_fleet_request.go#L610-L624), so there should be no need to add any extra logic in that sense. Just make sure the spot price & instance size is aligned with other requests from our testing suite. Feel free to take the "copy-paste-modify" approach if you like.
https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_spot_fleet_request_test.go

svenwltr commented:
@radeksimko AFAIS `active` only means, that the spot fleet request is created and enabled. It might still fail, if there is an error in the `launch_specification`.

I created [a test](https://github.com/rebuy-de/terraform-provider-aws/pull/1/commits/d9d85a4017e5639d6b5ad3eaa6b487595344c713) and removed my fix for testing. It succeeds, but it succeeded which shouldn't happen. This ist the History of the Spot Fleet Request:

![screenshot 2017-07-24 14 30 39 02 00](https://user-images.githubusercontent.com/1698599/28523457-cb50ccfa-707c-11e7-89a3-9d7a27227f49.png)

svenwltr commented:
Okay, this is what I found out after digging a bit deeper:

* A Spot Fleet Request has an `ActivityStatus` and a `SpotFleetRequestState` ([source](https://godoc.org/github.com/aws/aws-sdk-go/service/ec2#SpotFleetRequestConfig)). The "State" doesn't respect the progress of the "child instances".
* The [fulfillment check](https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_spot_fleet_request.go#L610-L624) uses `SpotFleetRequestState`, which means that the Terraform run might succeed, despite the actual instances don't start.
* If I use the `ActivityStatus` in the [fulfillment check](https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_spot_fleet_request.go#L610-L624) other tests also fail (or at least `TestAccAWSSpotFleetRequest_placementTenancy` with: `Non-Windows instances with a virtualization type of 'hvm' are currently not supported for this instance type.`)

It looks easy to change the behavior of the [fulfillment check](https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_spot_fleet_request.go#L610-L624), so it waits for the instances to actually be ready, but I cannot estimate the impact for such a change.

radeksimko commented:
Ah, I see, apologies for misleading info then!

It looks like we should add an optional fulfilment check, like we have in https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_spot_instance_request.go#L54-L58

Are you willing to do this in a separate PR?

> If I use the ActivityStatus in the fulfillment check other tests also fail (or at least TestAccAWSSpotFleetRequest_placementTenancy with: Non-Windows instances with a virtualization type of 'hvm' are currently not supported for this instance type.)

Yeah, that means those tests are _actually_ broken. 😞

svenwltr commented:
@radeksimko I guess it makes sense to add the fulfilment check to all existing test cases, right?

radeksimko commented:
@svenwltr pretty much. I first thought we could default to wait, but it seems the `aws_spot_instance` [**doesn't** wait by default](https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_spot_instance_request.go#L57), so better to stay consistent, which does mean we'll need to modify all existing test cases.

svenwltr commented:
Here is the implementation so far: terraform-providers/terraform-provider-aws#1241

I still need some time to fix the tests.

svenwltr commented:
@radeksimko I now rebased to the current master and fixed the `TestAccAWSSpotFleetRequest_withEBSDisk` test. Since this one already doesn't use a `key_name`, I didn't create an extra test. Is this sufficient?
